crates/tx5/src/peer/maybe_ready.rs

```diff
@@ -0,0 +1,207 @@
+use crate::backend::DynBackCon;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub(super) struct MaybeReady(Arc<Mutex<MaybeReadyState>>);
+
+enum MaybeReadyState {
+    Ready(DynBackCon),
+    Failed,
+    Wait(Arc<tokio::sync::Semaphore>),
+}
+
+impl MaybeReady {
+    pub(super) fn new() -> Self {
+        MaybeReady(Arc::new(Mutex::new(MaybeReadyState::Wait(Arc::new(
+            tokio::sync::Semaphore::new(0),
+        )))))
+    }
+
+    pub(super) fn query_ready<T>(
+        &self,
+        query: fn(&DynBackCon) -> T,
+    ) -> Option<T> {
+        let lock = self.0.lock().expect("poisoned");
+        match &*lock {
+            MaybeReadyState::Ready(c) => Some(query(c)),
+            _ => None,
+        }
+    }
+
+    pub(super) fn set_ready(&self, conn: DynBackCon) {
+        let mut lock = self.0.lock().expect("poisoned");
+        match &*lock {
+            MaybeReadyState::Wait(w) => {
+                // Close the wait semaphore to unblock waiters.
+                w.close();
+            }
+            MaybeReadyState::Failed => {
+                tracing::warn!(
+                    "Cannot set state to ready because state is already failed"
+                );
+                return;
+            }
+            MaybeReadyState::Ready(_) => {
+                tracing::error!(
```

Comment:
What's the reason for using `error!` here and `warn!` above in the `::Failed` case?